protocol/tx/entry.go

```diff
@@ -11,20 +11,44 @@ import (
 	"chain/protocol/bc"
 )
 
-type entry interface {
-	Type() string
-	Body() interface{}
-
-	// When an entry is created from a bc.TxInput or a bc.TxOutput, this
-	// reports the position of that antecedent object within its
-	// transaction. Both inputs (spends and issuances) and outputs
-	// (including retirements) are numbered beginning at zero. Entries
-	// not originating in this way report -1.
-	Ordinal() int
-}
+type (
+	entry interface {
+		Type() string
+		Body() interface{}
+
+		// When an entry is created from a bc.TxInput or a bc.TxOutput, this
+		// reports the position of that antecedent object within its
+		// transaction. Both inputs (spends and issuances) and outputs
+		// (including retirements) are numbered beginning at zero. Entries
+		// not originating in this way report -1.
+		Ordinal() int
+	}
+
+	// EntryRef holds an entry and its ID. The entry may be nil, but the
+	// ID of a valid entry must still be present.
+	EntryRef struct {
```

Comment:
I'd suggest not using the `type ( a; b )` form here. Rule of thumb for that is it's suitable for 3 or more one-liner type definitions in a row, or similar. IMO this is more readable as two separate statements.